src/mappings/handlers/pureProxyEventHandler.ts

```diff
@@ -3,30 +3,68 @@ import { SubstrateEvent } from "@subql/types";
 import { Proxied, PureProxy } from "../../types";
 import { extractProxyEventData, getProxiedId, getPureProxyId } from "../../utils/extractProxyEventData";
 
+import { calculatePureAccount } from "../../utils";
+import { extractPureProxyEventData } from "../../utils/extractPureProxyEventData";
+
 export async function handlePureProxyEvent(event: SubstrateEvent): Promise<void> {
-  const proxyData = extractProxyEventData(event);
+  const proxyData = extractPureProxyEventData(event);
 
   logger.info(`Pure Proxy Event: ${JSON.stringify(proxyData)}`);
   if (!proxyData) return;
 
-  const { proxyAccountId, accountId, type, delay, blockNumber, extrinsicIndex } = proxyData;
+  const { who, pure, type, disambiguationIndex, delay, blockNumber, extrinsicIndex } = proxyData;
+  let pureBlockNumber = blockNumber;
+
+  const typeString = type.toHuman() as string;
+  const typeU8a = type.toU8a();
+
+  const pureAccount = calculatePureAccount({
+    who: who,
+    proxyType: typeU8a,
+    index: disambiguationIndex,
+    maybeWhen: { blockHeight: blockNumber, extrinsicIndex },
+  });
+
+  if (pure !== pureAccount) {
+    const validationData = await api.query?.["parachainSystem"]?.["validationData"]?.();
+
+    if (!validationData) {
+      throw new Error(`Validation data on chain ${chainId} not found, time to die`);
+    }
+
+    const json = validationData?.toJSON() as { relayParentNumber: number };
+    const relayParentNumber = json?.relayParentNumber ?? 0;
+
+    const pureAccountRelayParent = calculatePureAccount({
+      who: who,
+      proxyType: typeU8a,
+      index: disambiguationIndex,
+      maybeWhen: { blockHeight: relayParentNumber, extrinsicIndex },
+    });
+
+    if (pure !== pureAccountRelayParent) {
+      throw new Error(`Who ${who} is not the pure account ${pureAccount} or the pure account relay parent ${pureAccountRelayParent}`);
+    }
+
+    pureBlockNumber = relayParentNumber;
+  }
 
   const pureProxy = PureProxy.create({
-    id: getPureProxyId({ chainId, accountId }),
+    id: getPureProxyId({ chainId, delegator: pure }),
```

Comment:
For `getPureProxyId` maybe we could name this field `pure` instead of `delegator`. `delegator` is right but tbh for me it always takes some additional mental power to proprely take into account `delegator` vs `delegatee` differences